From 5e6a835c1ee81bbf19ac759e428f07b582082d96 Mon Sep 17 00:00:00 2001
From: t-kuha <imagingtechnerd@gmail.com>
Date: Sat, 29 Mar 2025 12:00:18 +0000
Subject: [PATCH] Fix compilation error

---
 cmake/public/protobuf.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmake/public/protobuf.cmake b/cmake/public/protobuf.cmake
index 77ec3622b132..683dc07c4832 100644
--- a/cmake/public/protobuf.cmake
+++ b/cmake/public/protobuf.cmake
@@ -1,7 +1,7 @@
 # ---[ Protobuf
 
 # We will try to use the config mode first, and then manual find.
-find_package(Protobuf CONFIG QUIET)
+find_package(Protobuf QUIET)
 if(NOT Protobuf_FOUND)
   find_package(Protobuf MODULE QUIET)
 endif()
-- 
2.34.1

