diff --git a/tensorflow/lite/tools/make/download_dependencies.sh b/tensorflow/lite/tools/make/download_dependencies.sh
index 22ba383f688..e0394566ace 100755
--- a/tensorflow/lite/tools/make/download_dependencies.sh
+++ b/tensorflow/lite/tools/make/download_dependencies.sh
@@ -74,7 +74,7 @@ download_and_extract() {
   mkdir -p "${dir}"
   tempdir=$(mktemp -d)
   filepath="${tempdir}/$(basename ${url})"
-  curl -Lo ${filepath} ${url}
+  wget -q ${url} -O ${filepath} 
   if [ -n "${sha256}" ]; then
     echo "checking sha256 of ${dir}"
     echo "${sha256}  ${filepath}" | sha256sum -c
