From 0b1e3ff0529424aa388228a751a36d5de6026723 Mon Sep 17 00:00:00 2001
From: t-kuha <imagingnerd@gmail.com>
Date: Sat, 31 Jan 2026 21:29:22 +0900
Subject: [PATCH] Fix compile error

---
 CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index f886cf2e86..d329ccb91b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -47,7 +47,7 @@
 
 # TODO Lower to 3.24 when XNNPACK dependency is updated to include
 # https://github.com/google/XNNPACK/commit/c690daa67f883e1b627aadf7684c06797e9a0684
-cmake_minimum_required(VERSION 3.29)
+cmake_minimum_required(VERSION 3.28)
 project(executorch)
 
 set(EXECUTORCH_ROOT ${CMAKE_CURRENT_SOURCE_DIR})
-- 
2.43.0

